### PR TITLE
Supermarket, Rubocop, docs & fixes

### PR DESCRIPTION
**Pull request by afiune**
As a delivery user I need to have a way to resolve my cookbook dependencies
within my organization so I can develop cookbooks with multiple dependencies.

This PR Include the following commits:
* Setup a Supermarket Server
* Destroy Supermarket Server
* Render a better knife.rb using templates
* Updated description on metadata
* Bump metadata & small logic fix on Rakefile
* Extra documentation on README
* Rubocop Fixes

cc/ @opscode-cookbooks/delivery

### CODE REVIEW

**Comment by afiune** [comment]
Before:
````
salimafiune@afiuneChef:~/github/delivery-cluster
$ rubocop
warning: parser/current is loading parser/ruby21, which recognizes
warning: 2.1.6-compliant syntax, but you are running 2.1.5.
Inspecting 29 files
C.CCCWWCWCCC..WCWW.WCW.CWWCCC

29 files inspected, 343 offenses detected
````

After:
````
salimafiune@afiuneChef:~/github/delivery-cluster
$ rubocop
warning: parser/current is loading parser/ruby21, which recognizes
warning: 2.1.6-compliant syntax, but you are running 2.1.5.
Inspecting 30 files
.C...CCCCC....................

30 files inspected, 23 offenses detected
````

File: libraries/_helper.rb [hunk 1]
```diff
@@ -326,12 +386,31 @@ def upload_cookbook(cookbook)
...
     end
 
+    # Render a knife config file that points at the new delivery cluster
+    def render_knife_config
+      template File.join(cluster_data_dir, 'knife.rb') do
+        variables lazy {
+          {
+            chef_server_url:      chef_server_url,
+            client_key:           "#{cluster_data_dir}/delivery.pem",
+            analytics_server_url: if analytics_enabled?
```
Comment (christophermaier) [inline libraries/_helper.rb]: For the optional keys (`analytics_server_url` and `supermarket_site`) can we just not add them to the `knife.rb` at all?

Comment (afiune) [inline libraries/_helper.rb]: That's how it works.. :smile:

**Comment by jonsmorrow** [comment]
Overall this looks like a really goos enhancement. I am nervous that too many changes are coming in together. I'd like to see that cleanup bits such as changing to single quotes, whitespace fixes, and function name changes as a separate cleanup pr. Then bring in the new functionality. Both things have a lot of value but It is a bit confusing to understand the full delta.

**Comment by afiune** [comment]
@jonsmorrow I can tell you that this PR has been validated with both functionalities `SSH` and `AWS`.. Over the weekend I stud up both cluster twice (full automation with analytics, supermarket, splunk, everything) and I feel very confident that we can merge this change. But, if splitting the change makes you feel more confident, I can do that for you. :smile:

**Comment by jonsmorrow** [comment]
@afiune It is less about confidence and more about cleanliness of commit history if that makes sense. If it is too hard to split I'd say merge it. But in the future I think we should keep clean up changes separate from functionality changes.

File: README.md [hunk 2]
```diff
@@ -431,6 +462,10 @@ $ vi environments/kitchen.json
         "ip": "33.33.33.16",
         "username": "admin",
         "password": "salim"
+      },
+      "supermarket": {
+        "fqdn": "33.33.33.17",
+        "ip": "33.33.33.17"
```
Comment (jonsmorrow) [inline README.md]: Nothing to change in this pr. But we need to make fqdn honor ip or host if it is set. There's an issue filed from suncorp re this for delivery and chef-server: Issue #94

File: Rakefile [hunk 3]
```diff
@@ -37,66 +37,72 @@ Rake::TaskManager.record_task_metadata = true
...
-    msg "Create Delivery Server and Build Nodes"
+    msg 'Create Delivery Server and Build Nodes'
     chef_zero 'setup_delivery'
   end
 
   desc 'Activate Analytics Server'
-  task :analytics => [:chef_server] do
-    msg "Setup Chef Analytics so we can see what is going on in our cluster"
+  task analytics: [:chef_server] do
+    msg 'Setup Chef Analytics so we can see what is going on in our cluster'
```
Comment (jonsmorrow) [inline Rakefile]: I like the switch in syntax. But as my overall comment says I wish it were its own pr.

File: Rakefile [hunk 4]
```diff
@@ -37,66 +37,72 @@ Rake::TaskManager.record_task_metadata = true
...
+    system 'bundle exec berks update'
   end
 
   desc 'Clean the cache'
   task :clean_cache do
-    FileUtils.rm_rf(".chef/local-mode-cache")
-    FileUtils.rm_rf("cookbooks/")
+    FileUtils.rm_rf('.chef/local-mode-cache')
+    FileUtils.rm_rf('cookbooks/')
   end
```
Comment (jonsmorrow) [inline Rakefile]: I like the move to single quotes too, but wish it was in a clean up pr.

**Comment by afiune** [comment]
@jonsmorrow Uhmm I'm not sure I understand. If you are talking about commits, they have been separated, that's why I have 7 of them on this PR. I thought you was talking about the PR it self. If that is the case, I can put the commits on different PR's if you want. :smiley:

File: Rakefile [hunk 5]
```diff
@@ -135,38 +146,37 @@ end
...
-    system "cat .chef/delivery-cluster-data/*.creds"
+    # TODO: Use Ruby to read these files so its cross-platform
+    system 'cat .chef/delivery-cluster-data/*.creds'
   end
 
   desc 'List all your core services'
   task :list_core_services do
-    system "knife search node 'name:*server* OR name:build-node*' -a ipaddress"
-    system "grep chef_server_url .chef/delivery-cluster-data/knife.rb"
+    system 'knife search node \'name:*server* OR name:build-node*\' -a ipaddress'
```
Comment (jonsmorrow) [inline Rakefile]: This is one where switching to single quotes seems weird to me cause now we have to escape. I think for consisstency I want ingle quotes and escapes, but for readability I kinda want doubles on this.

File: libraries/_base.rb [hunk 6]
```diff
@@ -21,25 +21,23 @@
...
     # @author Salim Afiune <afiune@chef.io>
     class Base
-
       # Create a new Provisioning Driver Abstraction
       #
       # @param node [Chef::Node]
-      def initialize(node, run_context)
-        raise "#{self.class}#initialize must be implemented"
+      def initialize(node) # rubocop:disable Lint/UnusedMethodArgument
+        fail "#{self.class}#initialize must be implemented"
```
Comment (jonsmorrow) [inline libraries/_base.rb]: Why fail instead of raise? Is this just a style preference?

File: README.md (hunk 2 above)
Comment (afiune) [inline README.md]: Agree!

File: libraries/_base.rb (hunk 6 above)
Comment (afiune) [inline libraries/_base.rb]: Yes http://www.rubydoc.info/github/bbatsov/rubocop/master/Rubocop/Cop/Style/SignalException

File: Rakefile (hunk 3 above)
Comment (afiune) [inline Rakefile]: it is :smile:

File: Rakefile (hunk 4 above)
Comment (afiune) [inline Rakefile]: It is :smile:

File: libraries/_helper.rb [hunk 7]
```diff
@@ -173,7 +211,7 @@ def activate_splunk
       FileUtils.touch(splunk_lock_file)
     end
 
-    def is_splunk_enabled?
+    def splunk_enabled?
```
Comment (jonsmorrow) [inline libraries/_helper.rb]: sould prefer to see the function name refactor as it's own cleanup.

File: libraries/_helper.rb (hunk 1 above)
Comment (jonsmorrow) [inline libraries/_helper.rb]: Maybe put the conditional in a helper function? I think @christophermaier was suggesting we don't include the ke in the file unless we have a value. I tend to agree.

File: libraries/aws.rb [hunk 8]
```diff
@@ -91,10 +89,10 @@ def machine_options
       # @param component [String] component name
       # @param count [Integer] component number
```
Comment (jonsmorrow) [inline libraries/aws.rb]: You changed the name from count to _count comment should reflect this change.

File: libraries/_helper.rb [hunk 9]
```diff
@@ -326,12 +386,31 @@ def upload_cookbook(cookbook)
       end
     end
 
+    # Render a knife config file that points at the new delivery cluster
+    def render_knife_config
```
Comment (jonsmorrow) [inline libraries/_helper.rb]: I like this being pulled into a helper. good job.

File: vendor/chef-server-12/attributes/default.rb [hunk 10]
```diff
@@ -39,6 +39,9 @@
 # Analytics Server Parameters
 default['chef-server-12']['analytics'] = nil
 
+# Supermarket Server Parameters
+default['chef-server-12']['supermarket'] = nil
```
Comment (jonsmorrow) [inline vendor/chef-server-12/attributes/default.rb]: Did these also go back into 'chef-server-12' cookbook?

**Comment by jonsmorrow** [comment]
:+1: :shipit: :ship: All my comments are about style and ordering of changes. I think this would be hard to tease apart in this case and don't think we should block merging based on them.

File: vendor/chef-server-12/attributes/default.rb (hunk 10 above)
Comment (afiune) [inline vendor/chef-server-12/attributes/default.rb]: It did not.. Are we keeping that cookbook? cc/ @seth @christophermaier

File: libraries/_helper.rb (hunk 9 above)
Comment (afiune) [inline libraries/_helper.rb]: Thanks!! :smile_cat:

File: libraries/aws.rb (hunk 8 above)
Comment (afiune) [inline libraries/aws.rb]: I can fix that!

File: libraries/_helper.rb (hunk 7 above)
Comment (afiune) [inline libraries/_helper.rb]: That was part of the recomendations with Rubocop I think I can at least put it on the commit comment.. Thanks for the observation..

**Pull request merged**